examples/with-apollo/lib/apollo.js

```diff
@@ -1,22 +1,94 @@
 import React from 'react'
 import App from 'next/app'
-import Head from 'next/head'
 import { ApolloProvider } from '@apollo/react-hooks'
-import { ApolloClient } from 'apollo-client'
-import { InMemoryCache } from 'apollo-cache-inmemory'
-import { HttpLink } from 'apollo-link-http'
-import fetch from 'isomorphic-unfetch'
+import createApolloClient from '../apolloClient'
 
+// On the client we store the apollo client in the following variable
+// this prevents the client from reinitializing between page transitions.
 let globalApolloClient = null
 
 /**
- * Creates and provides the apolloContext
- * to a next.js PageTree. Use it by wrapping
- * your PageComponent via HOC pattern.
+ * Installes the apollo client on NextPageContext
+ * or NextAppContext. Useful if you want to use apolloClient
+ * inside getStaticProps, getStaticPaths or getServerProps
+ * @param {NextPageContext | NextAppContext} ctx
  */
-export const withApollo = ({ ssr = true } = {}) => PageComponent => {
+export const initOnContext = ctx => {
+  const inAppContext = Boolean(ctx.ctx)
+
+  // We consider installing `withApollo({ ssr: true })` on global App level
+  // as antipattern since it disables project wide Automatic Static Optimization.
+  if (process.env.NODE_ENV === 'development') {
+    if (inAppContext) {
+      console.warn(
+        'Warning: You have opted-out of Automatic Static Optimization due to `withApollo` in `pages/_app`.\n' +
+          'Read more: https://err.sh/next.js/opt-out-auto-static-optimization\n'
+      )
+    }
+  }
+
+  // Initialize ApolloClient if not already done
+  const apolloClient =
+    ctx.apolloClient ||
+    initApolloClient(ctx.apolloState || {}, inAppContext ? ctx.ctx : ctx)
+
+  // To avoid calling initApollo() twice in the server we send the Apollo Client as a prop
+  // to the component, otherwise the component would have to call initApollo() again but this
+  // time without the context, once that happens the following code will make sure we send
+  // the prop as `null` to the browser
+  apolloClient.toJSON = () => null
```

Comment:
Nice, I did this in next-with-apollo a long time ago, I don't remember why I never put it here 🤣

Reply:
Actually, is this required? I did it as a bug fix for an with configuration in the package, but I don't know about this case

Reply:
Reason is, it feels very hacky, the overhead of creating a new instance of the Apollo Client may not matter (unless it does, otherwise go ahead)

Reply:
I'm happy for this to be fixed in a follow-up! It's likely fixing an already existing bug.

Reply:
@lfades This line is there to prevent the following issue.

<img width="500" alt="Screenshot 2020-02-15 at 11 44 20" src="https://user-images.githubusercontent.com/1265681/74586531-a0fa8100-4fe8-11ea-8298-ded2976124fd.png">

I think it is a genius hack that reduces a lot of complexity inside this HOC.
I have to admit that this was not my idea. I believe I saw this pattern inside your excellent [lfades/next-with-apollo](https://github.com/lfades/next-with-apollo) module, the first time.